c/cert/src/rules/EXP39-C/DoNotAccessVariableViaPointerOfIncompatibleType.ql

```diff
@@ -0,0 +1,204 @@
+    // The memset call won't always have an edge to subsequent dereferences.
+    //
+    // Therefore, check that:
+    // 1) The memset call dominates the dereference.
+    // 2) The realloc call dominates the memset call.
+    // 3) There is no subsequent memset that also dominates the dereference.
+    //
+    // Currently, there is no relation between the pointer passed to memset
+    // and the pointer dereferenced. This unimplemented check might produce
+    // false-negatives when the memset call is unrelated to the reallocated memory.
+    not exists(FunctionCall memset, FunctionCall realloc, Expr ptr |
+      memset.getTarget() instanceof MemsetFunction and
+      realloc.getTarget() instanceof ReallocationFunction and
+      ptr = sink.asExpr() and
+      dominates(memset, ptr) and
+      not dominates(memset,
+        any(FunctionCall other |
+          other.getTarget() instanceof MemsetFunction and
+          other != memset and
+          dominates(other, ptr)
+        |
+          other
+        )) and
+      dominates(realloc, memset)
+    )
+  }
+
+  override predicate isBarrier(DataFlow::Node node, DataFlow::FlowState state) {
+    state instanceof IndirectCastReallocatedFlowState and
+    exists(FunctionCall fc |
+      fc.getTarget() instanceof MemsetFunction and
+      fc.getArgument(0) = node.asExpr()
+    )
+  }
+
+  override predicate isAdditionalFlowStep(
+    DataFlow::Node node1, DataFlow::FlowState state1, DataFlow::Node node2,
+    DataFlow::FlowState state2
+  ) {
+    // track pointer flow through realloc calls and update state to `IndirectCastReallocatedFlowState`
+    state1 instanceof IndirectCastDefaultFlowState and
+    state2 instanceof IndirectCastReallocatedFlowState and
+    exists(FunctionCall fc |
+      fc.getTarget() instanceof ReallocationFunction and
+      node1.asExpr() = fc.getArgument(fc.getTarget().(ReallocationFunction).getReallocPtrArg()) and
+      node2.asExpr() = fc
+    )
+    or
+    // track pointer flow through memset calls and reset state to `IndirectCastDefaultFlowState`
+    state1 instanceof IndirectCastReallocatedFlowState and
+    state2 instanceof IndirectCastDefaultFlowState and
+    exists(FunctionCall fc |
+      fc.getTarget() instanceof MemsetFunction and
+      node1.asExpr() = fc.getArgument(0) and
+      node2.asExpr() = fc
+    )
+  }
+}
+
+pragma[inline]
+predicate areTypesSameExceptForConstSpecifiers(Type a, Type b) {
+  a.stripType() = b.stripType() and
+  a.getSize() = b.getSize() and
+  forall(Specifier s | s = a.getASpecifier() and not s.hasName("const") |
+    b.hasSpecifier(s.getName())
+  )
+}
+
+pragma[inline]
+Type compatibleTypes(Type type) {
+  not (
+    type.isVolatile() and not result.isVolatile()
+    or
+    type.isConst() and not result.isConst()
+  ) and
+  (
+    (
+      result instanceof UnsignedCharType or
+      [result.stripTopLevelSpecifiers(), type.stripTopLevelSpecifiers()] instanceof VoidType
+    )
```

Comment:
Perhaps worth a comment on this case that it is a a deliberate cross-product? i.e. all types are associated with the unsigned char type.